datafusion/sql/src/unparser/plan.rs

```diff
@@ -455,19 +455,71 @@ impl Unparser<'_> {
             }
             LogicalPlan::SubqueryAlias(plan_alias) => {
                 let (plan, mut columns) =
-                    subquery_alias_inner_query_and_columns(&plan_alias);
+                    subquery_alias_inner_query_and_columns(plan_alias);
 
-                if !columns.is_empty() && !self.dialect.supports_column_alias_in_table_alias() {
-                    // if columns are returned than plan corresponds to a projection
-                    let LogicalPlan::Projection(inner_p) = plan else {
-                        return plan_err!(
-                            "Inner projection for subquery alias is expected"
-                        );
+                if !columns.is_empty()
+                    && !self.dialect.supports_column_alias_in_table_alias()
+                {
+                    let plan_rewrite_result = match plan {
+                        LogicalPlan::Projection(inner_p) => Ok(inject_column_aliases(inner_p, columns)),
+                        _ => {
+                            // complex subquery where projection is wrapped by other operator: LIMIT, SORT, DISTINCT, etc
+                            match plan.clone().map_children(|child| {
+                                if let LogicalPlan::Projection(p) = &child {
+                                    // Instead of specifying column aliases as part of the outer table, inject them directly into the inner projection
+                                    Ok(Transformed::yes(inject_column_aliases(p, columns.clone())))
+                                } else {
+                                    Ok(Transformed::no(child))
+                                }
+                            }) {
+                                Ok(plan) => Ok(plan.data),
+                                Err(e) => Err(e),
+                            }
+                        }
                     };
 
-                    // Instead of specifying column aliases as part of the outer table inject them directly into the inner projection
-                    let rewritten_plan = inject_column_aliases(&inner_p, &columns);
-                    columns.clear();
+                    let rewritten_plan = match plan_rewrite_result {
+                        Ok(p) => p,
+                        Err(e) => return internal_err!("Failed to transform SubqueryAlias plan: {e}")
+                    };
+                        // LogicalPlan::Limit(limit) => {
```

Comment:
This will be removed: just to show alternative approach with manual processing of plan variations instead of using `transform` functionality